commcare_connect/opportunity/visit_import.py

```diff
@@ -168,10 +169,11 @@ def update_payment_accrued(opportunity: Opportunity, users):
     """Updates payment accrued for completed and approved CompletedWork instances."""
     access_objects = OpportunityAccess.objects.filter(user__in=users, opportunity=opportunity, suspended=False)
     for access in access_objects:
-        completed_works = access.completedwork_set.exclude(
-            status__in=[CompletedWorkStatus.rejected, CompletedWorkStatus.over_limit]
-        ).select_related("payment_unit")
-        update_status(completed_works, access, True)
+        with cache.lock(f"update_payment_accrued_lock_{access.id}"):
```

Comment:
how does this handle failure? when does the lock timeout to avoid deadlocks, and if multiple forms come in quick succession will this block the for receiver view for an extended period?

Reply:
This cache lock has blocking behavior, so if requests come in quick succession they will be blocked until the code inside the lock has completed executing. The timeouts on these locks are infinite by default i.e. they wait for the execution of code to complete before releasing the lock.

Reply:
We should add a timeout to avoid deadlock. Longer than the work will actually take, but that way if there is an unhandled exception or unexpected failure, it wont block all other work forever

Reply:
Fixed. [d33b0af](https://github.com/dimagi/commcare-connect/pull/416/commits/d33b0af2d8a452dab6068aad6a37da20883a7acb)